src/connection.js

```diff
@@ -1462,74 +1343,31 @@ Connection.prototype.STATE = {
         return this.transitionTo(this.STATE.FINAL);
       },
       message: function() {
-        return this.processLogin7Response();
-      }
-    }
-  },
-  SENT_LOGIN7_WITH_NTLM: {
-    name: 'SentLogin7WithNTLMLogin',
-    events: {
-      socketError: function() {
-        return this.transitionTo(this.STATE.FINAL);
-      },
-      connectTimeout: function() {
-        return this.transitionTo(this.STATE.FINAL);
-      },
-      data: function(data) {
-        if (this.sspiClientResponsePending) {
-          // We got data from the server while we're waiting for getNextBlob()
-          // call to complete on the client. We cannot process server data
-          // until this call completes as the state can change on completion of
-          // the call. Queue it for later.
-          const boundDispatchEvent = this.dispatchEvent.bind(this);
-          return setImmediate(boundDispatchEvent, 'data', data);
-        } else {
-          return this.sendDataToTokenStreamParser(data);
+        // Use SSPI blob if received
+        if (this.sspiBuffer) {
+          return this.authProvider.handshake(this.sspiBuffer, (error, responseBuffer) => {
+            if (error) {
+              this.emit('error', error);
+              return this.close();
+            }
+            this.messageIo.sendMessage(TYPE.NTLMAUTH_PKT, responseBuffer);
+
```

Comment:
Send `NTLMAUTH_PKT` only if `responseBuffer` is valid, as not all auth mechanism require this packet. For instance Kerberos handshake is complete in one round trip and doesn't involve `NTLMAUTH_PKT`.

Reply:
During kerberos auth, does the server send back a SSPI token? 🤔

If not, then this code should be okay, as it only sends another `NTLMAUTH_PKT` in case a SSPI token was received. 🤷‍♂️

Reply:
Server does send SSPI token for Kerberos. Unlike NTLM and Windows integrated auth, Client/Server validation is complete in one round-trip for Kerberos.
 
TDS document in section `2.2.6.8`  says `The client MUST respond with another SSPI message`, but it is not applicable for Kerberos, as it contradicts the handshake explained in RFC4120 😕. You can refer to exchange `(5)` and `(6)` of handshake diagram in https://msdn.microsoft.com/en-us/library/cc233863.aspx, for extended reading you can check section `3.2` in [RFC4120](https://www.ietf.org/rfc/rfc4120.txt).

Reply:
@v-suhame Unfortunately, I can't double check the actual behaviour under kerberos, as I don't have a test environment with domain support set up (and kerberos does not support workstation based login). 😞 

Could you double check this that the server really sends a `SSPI` token during kerberos authentication?

Reply:
No worries :) 
In Kerberos, after sending `Login7` packet, packet with `SSPI` token and a packet with `ENVCHANGE` token are received one after another. This avoids the need for `NTLMAUTH_PKT` packet as Server has already authenticated the Client.

Reply:
Oh, okay. 🤔 I'll fix that up then! 👍